File: src/neetcode/Java/_32LongestValidParentheses.java
package neetcode.Java;
import java.util.Stack;

public class _32LongestValidParentheses {

    public int longestValidParentheses(String s) {
        Stack<Integer> st = new Stack<>();
        st.add(-1);
        int max = 0;
        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) == '(')
                st.add(i);
            else {
                st.pop();
                if (st.isEmpty())
                    st.add(i);
                else
                    max = Math.max(max, i - st.peek());
            }
        }
        return max;
    }

    public int longestValidParentheses1(String s) {
        Stack<Integer> st = new Stack<>();
        st.push(-1);
        int max = 0;
        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) == '(')
                st.push(i);
            else {
                st.pop();
                if (!st.isEmpty())
                    max = Math.max(max, i - st.peek());
                else
                    st.push(i);
            }
        }
        return max;

    }

    public static void main(String[] args) {
        _32LongestValidParentheses obj = new _32LongestValidParentheses();
        String s = ")()(())";
        System.out.println(obj.longestValidParentheses(s));
        System.out.println(obj.longestValidParentheses1(s));
    }
}

File: src/neetcode/kotlin/_32LongestValidParentheses.kt
package neetcode.kotlin

import java.util.*
import kotlin.math.max

class _32LongestValidParentheses {
    fun longestValidParentheses(s: String): Int {
        val st = Stack<Int>()
        st.add(-1)
        var max = 0
        for (i in 0 until s.length) {
            if (s[i] == '(') st.add(i)
            else {
                st.pop()
                if (st.isEmpty()) st.add(i)
                else max = max(max.toDouble(), (i - st.peek()).toDouble()).toInt()
            }
        }
        return max
    }

    fun longestValidParentheses1(s: String): Int {
        val st = Stack<Int>()
        st.push(-1)
        var max = 0
        for (i in 0 until s.length) {
            if (s[i] == '(') st.push(i)
            else {
                st.pop()
                if (!st.isEmpty()) max = max(max.toDouble(), (i - st.peek()).toDouble()).toInt()
                else st.push(i)
            }
        }
        return max
    }

    companion object {
        @JvmStatic
        fun main(args: Array<String>) {
            val obj = _32LongestValidParentheses()
            val s = ")()(())"
            println(obj.longestValidParentheses(s))
            println(obj.longestValidParentheses1(s))
        }
    }
}
